phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -823,6 +844,42 @@ private boolean verifySingleIndexRow(Result indexRow, VerificationResult.PhaseRe
             if (actualMutationIndex == actualMutationList.size()) {
                 break;
             }
+            // Skip newer expected unverified put or delete mutations
+            if (getTimestamp(expectedMutation) > getTimestamp(actualMutation) &&
+                    ((expectedMutation instanceof Put && !isVerified(expectedMutation) && expectedMutation.size() == 1) ||
+                            (expectedMutation instanceof Delete))) {
+                // The reason there is no corresponding actual index row mutations is because of concurrent data
+                // table mutations. For each data table mutation, two operations are done on the data table. One is to
+                // read the existing row state, and the second is to write to the data table. The processing of
+                // concurrent data mutations are serialized once for reading the existing row states, and then
+                // serialized again for updating data table. In other words, they go through locking twice, i.e.,
+                // [lock, read, unlock] and [lock, write, unlock]. Because of this two phase locking, for a pair of
+                // concurrent mutations(for the same row), the same row state is read from the data table. This means
+                // the same existing index row can be made unverified twice with different timestamps, one for each
+                // the concurrent mutation. So skipping newer expected unverified put mutation are safe. One of these
+                // unverified put mutations is later cleaned up by delete markers by the read repair process.
+                // We can skipped these delete markers safely above. When the data table rows are rebuild, the rebuild
+                // process generates the delete family markers. The timestamp of delete markers are the timestamp of
+                // the data table mutation for which the delete marker is added. This means their timestamp will be
+                // higher than the timestamp of actual index row to be deleted.
+                continue;
+            }
+            // For each expected delete (family) mutation there has to be the corresponding actual unverified put
+            // mutation. The corresponding delete mutation may not be there if the last write phase does not
+            // make it. The following two patterns are possible
+            // {(expected delete, actual delete), (expected unverified put, actual unverified put)},
+            // {(expected delete, actual unverified put), (expected unverified put, no corresponding actual put)}
+            if (expectedMutation instanceof Delete && isDeleteFamily(expectedMutation) &&
```

Comment:
shouldn't this be isDeleteFamilyVersion, after PHOENIX-5743?

Reply:
A DeleteFamilyVersion delete marker is only added by the read repair part implemented by GlobalIndexChecker. It is added only if the unverified put is a new version of an existing index row. The (single row or all row) rebuild adds delete markers only of type DeleteFamily.